### PR TITLE
OCPQE-15785: registries containers definitions

### PR DESCRIPTION
**Pull request by zniu1011**

### CODE REVIEW

**Review by aleskandro — COMMENTED** [review]
Hi @zniu1011, thanks! Adding some comments.

**Comment by zniu1011** [comment]
Thanks for the comments @aleskandro, let me check and update.

**Comment by zniu1011** [comment]
@aleskandro I made some updates, and have 3 instances conf now, we can start the service by "systemctl start poc-registry@registry${PORT}.service".
There is no data to content  /opt/registry-${port}/data for now, for the sensitive datas (auth and certs), I made some fake dates to content during the test.

**Comment by aleskandro** [comment]
> @aleskandro I made some updates, and have 3 instance confs now, we can start the service by "systemctl start poc-registry@registry${PORT}.service". There is no data to content /opt/registry-${port}/data for now, for the sensitive data (auth and certs), I made some fake data to content during the test.

Hi @zniu1011 is it possible to add the three instances we need to run in the fcos-bastion-image/root/usr/lib/systemd/system-preset/98-bastion.preset

**Comment by zniu1011** [comment]
> > @aleskandro I made some updates, and have 3 instance confs now, we can start the service by "systemctl start poc-registry@registry${PORT}.service". There is no data to content /opt/registry-${port}/data for now, for the sensitive data (auth and certs), I made some fake data to content during the test.
> 
> Hi @zniu1011 is it possible to add the three instances we need to run as enabled by default in the fcos-bastion-image/root/usr/lib/systemd/system-preset/98-bastion.preset
> 
> See this file as an example: https://github.com/openshift-qe/baremetal-qe-infra/blob/471cdf99471cfac9d3cdd6f69c5cbe24e288b089/images/fcos-base-image/root/usr/lib/systemd/system-preset/97-fedora-custom.preset

Thanks @aleskandro , added.

File: images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    path: /auth/htpasswd
+health:
+  storagedriver:
+    enabled: true
+    interval: 10s
+    threshold: 3
+EOF
+
+## REGISTRY PASSWORD
+echo 'dummy:$2y$05$7RzcQXdIGYLc6hI0luBGVuQGt2CB1WjFaX48JVgKBAt.BzSMZ/Q6G' > "${registry_password_file}"
```
Comment (aleskandro) [inline images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh:42]: Can we consider this similarly to reg_cert_file? Assume you will have an htpasswd file in `${htpasswd_file}` (for example /opt/registry-common/htpasswd), you can copy that to `${registry_password_file}`.

The files in /opt/registry-common/ will be provided externally and the unit will be able to start and do the initial configuration.


Also add


```
ConditionPathExists=/opt/registry-common
```
in the `[unit]` section of the unit file: https://github.com/openshift-qe/baremetal-qe-infra/pull/13/files#diff-0471a597017ffe2f8d3aa15a1c6507a2e8b83dbdb703b2a959d1570bebee5ac2R5

File: images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#!/bin/bash
```
Comment (aleskandro) [inline images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh:1]: this file should stay at `root/usr/bin` IMO

File: images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    interval: 10s
+    threshold: 3
+EOF
+
+## REGISTRY PASSWORD
+echo 'dummy:$2y$05$7RzcQXdIGYLc6hI0luBGVuQGt2CB1WjFaX48JVgKBAt.BzSMZ/Q6G' > "${registry_password_file}"
+
+## CERTS FILES
+cat "${reg_cert_file}" > "${registry_certs_cert}"
+cat "${reg_key_file}"> "${registry_certs_key}"
```
Comment (aleskandro) [inline images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh:46]:
```suggestion
cat "${reg_key_file}"> "${registry_certs_key}"
touch /var/.registry-$PORT

```

File: images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  storagedriver:
+    enabled: true
+    interval: 10s
+    threshold: 3
+EOF
+
+## REGISTRY PASSWORD
+echo 'dummy:$2y$05$7RzcQXdIGYLc6hI0luBGVuQGt2CB1WjFaX48JVgKBAt.BzSMZ/Q6G' > "${registry_password_file}"
+
+## CERTS FILES
```
Comment (aleskandro) [inline images/fcos-base-image/root/usr/share/bastion_services/registry/gen_registry_file.sh:44]:
```suggestion
## CERTS FILES
if [ ! -f "${reg_cert_file}" ] || [ ! -f "${reg_key_file}" ]; then
  echo "Some files are not ready yet. Exiting
  exit 1
fi
```

File: images/fcos-base-image/root/usr/lib/systemd/system/poc-registry@.service [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+Type=simple
+TimeoutStartSec=70
+ExecStartPre=/usr/bin/podman rm -f poc-registry-${PORT}
+ExecStartPre=/bin/sh -c '/usr/bin/mkdir -p /opt/registry-${PORT}/{data,auth,certs}'
+ExecStartPre=/bin/sh -c '/usr/bin/gen_registry_file.sh ${PORT}'
+ExecStart=/usr/bin/podman run --rm --name poc-registry-${PORT} \
+                              -p ${PORT}:${PORT} \
+                              --net host \
+                              -v /opt/registry-${PORT}/data:/var/lib/registry:z \
+                              -v /opt/registry-${PORT}/auth:/auth \
```
Comment (aleskandro) [inline images/fcos-base-image/root/usr/lib/systemd/system/poc-registry@.service:21]:
```suggestion
                              -v /opt/registry-${PORT}/auth:/auth:Z \
```

File: images/fcos-base-image/root/usr/lib/systemd/system/poc-registry@.service [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+                              -v /opt/registry-${PORT}/certs:/certs:z \
+                              -v /opt/registry-${PORT}/config.yaml:/etc/docker/registry/config.yml \
```
Comment (aleskandro) [inline images/fcos-base-image/root/usr/lib/systemd/system/poc-registry@.service:23]:
```suggestion
                              -v /opt/registry-${PORT}/config.yaml:/etc/docker/registry/config.yml:Z \
```

**Review by aleskandro — COMMENTED** [review]
Hi @zniu1011, it generally looks good to me, just suggesting a few other fixes for nits.

Have you tested this? 

Also rebase to the latest changes and squash the commits before merging.

**Comment by zniu1011** [comment]
> Hi @zniu1011, it generally looks good to me, just suggesting a few other fixes for nits.
> 
> Have you tested this?
> 
> Also rebase to the latest changes and squash the commits before merging.

@aleskandro , I tested on a server, it works good, when there is no registry-common directory, or no required certs files in the directory, there will show error in 'systemctl status poc-registry@registry-${PORT}" or in /var/log/message.

**Comment by aleskandro** [comment]
ok, if you also commit https://github.com/openshift-qe/baremetal-qe-infra/pull/13#discussion_r1309851353 and squash, we can merge this.

File: images/fcos-base-image/root/usr/lib/systemd/system/poc-registry@.service [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+EnvironmentFile=/usr/share/bastion_services/registry/%i.conf
+Restart=on-failure
+Type=simple
+TimeoutStartSec=70
+ExecStartPre=/usr/bin/podman rm -f poc-registry-${PORT}
+ExecStartPre=/bin/sh -c '/usr/bin/mkdir -p /opt/registry-${PORT}/{data,auth,certs}'
+ExecStartPre=/bin/sh -c '/usr/bin/gen_registry_file.sh ${PORT}'
+ExecStart=/usr/bin/podman run --rm --name poc-registry-${PORT} \
+                              -p ${PORT}:${PORT} \
+                              --net host \
```
[the remark this answers is not in the document]
Comment (aleskandro) [reply images/fcos-base-image/root/usr/lib/systemd/system/poc-registry@.service:18]: Is there a reason not to make this change?

**Comment by zniu1011** [comment]
> /lgtm cancel
Sorry about that @aleskandro , I am commit comment #13 now.

**Comment by aleskandro** [comment]
we miss another thing @zniu1011 . These files should stay in the `fcos-bastion-image` folder, not the `fcos-base-image` one.

**Comment by zniu1011** [comment]
Reopen to push the new update.

**Review by aleskandro — APPROVED** [review]
/hold cancel

Thanks @zniu1011. I apologize for the back and forth. I didn't see the folder you were using early in the reviews.

**Comment by zniu1011** [comment]
I updated the location mistake. Thanks @aleskandro, for the review and supports.

**Pull request merged**